### PR TITLE
[Snyk] Security upgrade transformers from 4.53.0 to 5.3.0

### PR DESCRIPTION
**Pull request by karly-quick**
snyk-top-banner

#### Snyk changed the following file(s):

- `models/tt_transformers/requirements.txt`

#### Breaking Change Risk

Merge Risk: High

> **Notice:** This assessment is enhanced by AI.

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/tt-test-rust/project/1b5e0150-313e-44f7-b854-5865b605407b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/tt-test-rust/project/1b5e0150-313e-44f7-b854-5865b605407b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Deserialization of Untrusted Data](https://learn.snyk.io/lesson/insecure-deserialization/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":"high","FF_showPullRequestBreakingChanges":true,"FF_showPullRequestBreakingChangesWebSearch":false,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"transformers","from":"4.53.0","to":"5.3.0"}],"env":"prod","issuesToFix":["SNYK-PYTHON-TRANSFORMERS-16876613"],"prId":"073354b6-6e70-4ea4-b462-ee769249eedb","prPublicId":"073354b6-6e70-4ea4-b462-ee769249eedb","packageManager":"pip","priorityScoreList":[148],"projectPublicId":"1b5e0150-313e-44f7-b854-5865b605407b","projectUrl":"https://app.snyk.io/org/tt-test-rust/project/1b5e0150-313e-44f7-b854-5865b605407b?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-TRANSFORMERS-16876613"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by karly-quick** [comment]
Merge Risk: High

The upgrade from `transformers` v4 to v5 is a **high-risk** major update with significant and numerous breaking changes. It is the first major release in five years and modernizes the entire library, but requires careful migration.  
  
**Key Breaking Changes:**  
  
*   **PyTorch-Only Backend:** Support for TensorFlow and Flax has been completely removed. [2, 4] All `TF*` and `Flax*` models and classes are gone. The library is now exclusively based on PyTorch. [2]  
*   **Unified Tokenizer System:** The distinction between "slow" (Python) and "fast" (Rust) tokenizers has been eliminated. [1, 3] A single, unified API now automatically selects the optimal backend, but this refactor removes old import paths and methods like `encode_plus`. [1]  
*   **Quantization API Overhaul:** The `load_in_4bit` and `load_in_8bit` arguments for `from_pretrained` have been removed. [3] All quantization is now handled through a new, unified `quantization_config` object. [1]  
*   **Model Loading Behavior:** The `dtype` parameter in `from_pretrained` now defaults to `"auto"`, meaning models will load in their native precision (e.g., float16) instead of defaulting to float32. This can cause silent numerical differences if not handled explicitly. [4]  
*   **Removed Features:** Several less-used features have been removed, including head masking, pruning, and relative positional biases in some models. [1, 3, 6]  
*   **CLI Changes:** The main command-line interface is now `transformers`, replacing the old `transformers-cli`. The `transformers chat` command has also been significantly refactored. [1, 8]  
  
**Recommendation:**  
This upgrade will almost certainly require code changes. Do not merge without a thorough migration effort. Developers must validate their entire workflow, from data processing and tokenization to model loading and training/inference logic. It is critical to review the official migration guide provided by Hugging Face before attempting the upgrade.  
  
**Source:** [Official Migration Guide](https://github.com/huggingface/transformers/blob/main/MIGRATION_GUIDE_V5.md) [3], [Release Announcements](https://huggingface.co/docs/transformers/v5.0.0/en/blog/2025-12-01-transformers-v5) [2, 8]

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.

**Comment by karly-quick** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/pr-checks/d3ed8eca-c5dd-4ee4-9930-959430e683c7?refs=cfc11df9ab12c160d099ae5a2009d6b3a0b672c3&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/pr-checks/d3ed8eca-c5dd-4ee4-9930-959430e683c7/license?refs=cfc11df9ab12c160d099ae5a2009d6b3a0b672c3&source=prChecksComment) |
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/project/c45c259d-ddf0-4be1-960e-6f383f9dbef2/pr-check/5c018cf6-d6dd-4b51-989d-20f199459e8c?refs=cfc11df9ab12c160d099ae5a2009d6b3a0b672c3&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).
